janitor/functions.py

```diff
@@ -3869,3 +3813,230 @@ def sort_naturally(
     """
     new_order = index_natsorted(df[column_name], **natsorted_kwargs)
     return df.iloc[new_order, :]
+
+
+@pf.register_dataframe_method
+def expand_grid(df: pd.DataFrame = None,
+                df_key: str = None,
+                others: Dict = None) -> pd.DataFrame:
+    """ Creates a dataframe from all combinations of all inputs.
+        This method mutates the original dataframe.
+
+    Functional usage syntax:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor as jn
+
+        df = pd.DataFrame(...)
+        df = jn.expand_grid(df=df, df_key="...", others={...})
+
+    Method-chaining usage syntax:
+
+        import pandas as pd
+        import janitor as jn
+
+    .. code-block:: python
+
+        df = pd.DataFrame(...).expand_grid(df_key="bla",others={...})
+
+    :param df: A pandas dataframe.
+    :param df_key: name of key for the dataframe. It becomes the column name of the dataframe.
+    :others: A dictionary. This will contain the data to be combined with the dataframe.
+    :returns : A pandas dataframe of all combinations of name value pairs.
+
+    This works with a dictionary of name value pairs, and will work with structures that are not dataframes.
+    The output will always be a dataframe.
```

Comment:
Two changes requested please!

Firstly, can we move this above the parameters section?

Secondly, can you apply [semantic line breaks](http://sembr.org/) here?

```suggestion
    This works with a dictionary of name value pairs, 
    and will work with structures that are not dataframes.
    The output will always be a dataframe.
```